source/qnx/ionotify_event_loop.c

```diff
@@ -0,0 +1,1042 @@
+    int *errno_value) {
+    /* Event of type SIGEV_UNBLOCK makes the timed-out kernel call fail with an error of ETIMEDOUT. */
+    struct sigevent notify;
+    SIGEV_UNBLOCK_INIT(&notify);
+    int rc = TimerTimeout(CLOCK_MONOTONIC, _NTO_TIMEOUT_RECEIVE, &notify, timeout, NULL);
+    if (rc == -1) {
+        *errno_value = errno;
+        return rc;
+    }
+    rcvid_t rcvid = MsgReceive(io_events_channel_id, pulse, sizeof(*pulse), NULL);
+    if (rcvid == -1) {
+        *errno_value = errno;
+    }
+    return rcvid;
+}
+
+static void s_aws_ionotify_cleanup_aws_lc_thread_local_state(void *user_data) {
+    (void)user_data;
+    aws_cal_thread_clean_up();
+}
+
+static void s_process_pulse(
+    struct aws_event_loop *event_loop,
+    const struct _pulse *pulse,
+    bool *should_process_cross_thread_tasks) {
+    if (pulse->code == CROSS_THREAD_PULSE_SIGEV_CODE) {
+        AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: MsgReceive got cross-thread pulse", (void *)event_loop);
+        *should_process_cross_thread_tasks = true;
+        return;
+    }
+
+    int user_data = pulse->value.sival_int;
+
+    int handle_id = user_data & _NOTIFY_DATA_MASK;
+    if (handle_id == 0) {
+        AWS_LOGF_ERROR(AWS_LS_IO_EVENT_LOOP, "id=%p: Got pulse with empty handle ID, ignoring it", (void *)event_loop);
+        return;
+    }
+
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Got pulse for handle ID %u", (void *)event_loop, handle_id);
+
+    struct ionotify_loop *ionotify_loop = event_loop->impl_data;
+    struct ionotify_event_data *ionotify_event_data = s_find_handle(event_loop, ionotify_loop, handle_id);
+    if (ionotify_event_data == NULL) {
+        /* This situation is totally OK when the corresponding fd is already unsubscribed. */
+        AWS_LOGF_DEBUG(
+            AWS_LS_IO_EVENT_LOOP,
+            "id=%p: No mapped data found for handle ID %d, fd must be already unsubscribed",
+            (void *)event_loop,
+            handle_id);
+        return;
+    }
+
+    if (!ionotify_event_data->is_subscribed) {
+        return;
+    }
+
+    AWS_LOGF_TRACE(
+        AWS_LS_IO_EVENT_LOOP,
+        "id=%p: Processing fd %d: pulse code %d",
+        (void *)event_loop,
+        ionotify_event_data->handle->data.fd,
+        pulse->code);
+    int event_mask = 0;
+    if (pulse->value.sival_int & _NOTIFY_COND_OBAND) {
+        AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: fd got out-of-band data", (void *)event_loop);
+        event_mask |= AWS_IO_EVENT_TYPE_READABLE;
+    }
+    if (pulse->value.sival_int & _NOTIFY_COND_INPUT) {
+        AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: fd is readable", (void *)event_loop);
+        event_mask |= AWS_IO_EVENT_TYPE_READABLE;
+    }
+    if (pulse->value.sival_int & _NOTIFY_COND_OUTPUT) {
+        AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: fd is writable", (void *)event_loop);
+        event_mask |= AWS_IO_EVENT_TYPE_WRITABLE;
+    }
+    if (pulse->value.sival_int & _NOTIFY_COND_EXTEN) {
```

Comment:
Under what circumstances would this happen and what does it mean?  All we seem to do is log it.

Reply:
Oh, this one is tough. I added a comment describing my interaction with this flag. I suggest we keep this log, maybe it'll reveal something.